interfaces/apparmor/backend.go

```diff
@@ -0,0 +1,175 @@
+	if err != nil {
+		return fmt.Errorf("cannot obtain security snippets for snap %q: %s", snapInfo.Name, err)
+	}
+	// Get the files that this snap should have
+	content, err := b.CombineSnippets(snapInfo, developerMode, snippets)
+	if err != nil {
+		return fmt.Errorf("cannot obtain expected security files for snap %q: %s", snapInfo.Name, err)
+	}
+	changed, removed, err := osutil.EnsureDirState(b.Directory(), b.FileGlob(snapInfo), content)
+	if err != nil {
+		return fmt.Errorf("cannot synchronize security files for snap %q: %s", snapInfo.Name, err)
+	}
+	err = b.ObserveChanges(changed, removed)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Deconfigure removes security artefacts of a given snap.
+func (b *Backend) Deconfigure(snapInfo *snap.Info) error {
+	_, removed, err := osutil.EnsureDirState(b.Directory(), b.FileGlob(snapInfo), nil)
+	if err != nil {
+		return fmt.Errorf("cannot synchronize security files for snap %q: %s", snapInfo.Name, err)
+	}
+	err = b.ObserveChanges(nil, removed)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// CombineSnippets combines security snippets collected from all the interfaces
+// affecting a given snap into a content map applicable to EnsureDirState. The
+// backend delegates writing those files to higher layers.
+func (b *Backend) CombineSnippets(snapInfo *snap.Info, developerMode bool, snippets map[string][][]byte) (content map[string]*osutil.FileState, err error) {
+	for _, appInfo := range snapInfo.Apps {
+		s := make([][]byte, 0, len(snippets[appInfo.Name])+2)
+		s = append(s, b.aaHeader(appInfo, developerMode))
+		s = append(s, snippets[appInfo.Name]...)
+		s = append(s, []byte("}\n"))
+		fileContent := bytes.Join(s, []byte("\n"))
+		if content == nil {
+			content = make(map[string]*osutil.FileState)
+		}
+		content[b.FileName(appInfo)] = &osutil.FileState{
+			Content: fileContent,
+			Mode:    0644,
+		}
+	}
+	return content, nil
+}
+
+// ObserveChanges informs the backend about changes made to the set of managed
+// files by a higher layer.
+//
+// The backend may choose to react to those changes immediately or to buffer
+// them until the higher layer signals that no more changes are coming by
+// calling FinishChanges.
+//
+// Buffering the changes is desirable when a constant cost can be incurred
+// regardless of the number of changes made.
+func (b *Backend) ObserveChanges(changed, removed []string) error {
+	// Reload changed profiles
+	for _, baseName := range changed {
+		fname := filepath.Join(b.Directory(), baseName)
+		err := LoadProfile(fname)
+		if err != nil {
+			return fmt.Errorf("cannot load apparmor profile %q: %s", baseName, err)
+		}
+	}
+	// Unload removed profiles
+	for _, baseName := range removed {
+		// XXX: This depends on FileName() and ProfileName() returning the same
+		// value. This is true since both of those use SecurityTag().
+		profile := Profile{Name: baseName}
+		err := profile.Unload()
```

Comment:
This part needs some cleanup/consideration of how/if we want to scan all loaded apparmor profiles. Earlier it seemed we do want to scan loaded profiles. This gave rise of the Profile struct. Now I'm not so sure. The only point of that struct is to convey profile name _and_ enforcement mode together. If we don't need this information the API can be simplified.